File: langs/dotnet/NobleIDL.Tests/DefaultValueTests.cs
using System.Numerics;
using ESExpr.Runtime;
using NUnit.Framework.Legacy;

namespace NobleIDL.Tests;


public class DefaultValueTests {

    [Test]
    public void CheckDefaultValues() {
        var vExpr = new Expr.Constructor("default-values", [], default);
        IESExprCodec<DefaultValues> codec = new DefaultValues.Codec();
        var v = codec.Decode(vExpr);

        Assert.Multiple(() => {
            Assert.That(v.BoolTrue, Is.True);
            Assert.That(v.BoolFalse, Is.False);
            Assert.That(v.StrValue, Is.EqualTo("abc"));
            Assert.That(v.BinaryValue.ByteArray, Is.EqualTo(new byte[] { 0xAB, 0xCD, 0xEF }));
            
            Assert.That(v.IntValue55, Is.EqualTo((BigInteger)55));
            Assert.That(v.IntBeyondInt64, Is.EqualTo(BigInteger.Parse("18446744073709551616")));
            Assert.That(v.IntMinus55, Is.EqualTo((BigInteger)(-55)));
            Assert.That(v.IntBeyondNegInt64, Is.EqualTo(BigInteger.Parse("-18446744073709551617")));
            Assert.That(v.NatValue55.BigIntegerValue, Is.EqualTo((BigInteger)55));
            Assert.That(v.NatBeyondInt64.BigIntegerValue, Is.EqualTo(BigInteger.Parse("18446744073709551616")));
            
            Assert.That(v.I8Value55, Is.EqualTo((sbyte)55));
            Assert.That(v.I8Minus55, Is.EqualTo((sbyte)(-55)));
            Assert.That(v.U8Value55, Is.EqualTo((byte)55));
            Assert.That(v.U8Value255, Is.EqualTo(byte.MaxValue));
            Assert.That(v.I16Value55, Is.EqualTo((short)55));
            Assert.That(v.I16Minus55, Is.EqualTo((short)(-55)));
            Assert.That(v.U16Value55, Is.EqualTo((ushort)55));
            Assert.That(v.U16ValueMax, Is.EqualTo(ushort.MaxValue));
            Assert.That(v.I32Value55, Is.EqualTo(55));
            Assert.That(v.I32Minus55, Is.EqualTo(-55));
            Assert.That(v.U32Value55, Is.EqualTo(55U));
            Assert.That(v.U32ValueMax, Is.EqualTo(uint.MaxValue));
            Assert.That(v.I64Value55, Is.EqualTo(55L));
            Assert.That(v.I64Minus55, Is.EqualTo(-55L));
            Assert.That(v.U64Value55, Is.EqualTo(55UL));
            Assert.That(v.U64ValueMax, Is.EqualTo(ulong.MaxValue));
            
            Assert.That(v.F32Value55, Is.EqualTo(55.0f));
            Assert.That(v.F32Minus55, Is.EqualTo(-55.0f));
            Assert.That(v.F32Nan, Is.NaN);
            Assert.That(v.F32Inf, Is.EqualTo(float.PositiveInfinity));
            Assert.That(v.F32MinusInf, Is.EqualTo(float.NegativeInfinity));
            
            Assert.That(v.F64Value55, Is.EqualTo(55.0));
            Assert.That(v.F64Minus55, Is.EqualTo(-55.0));
            Assert.That(v.F64Nan, Is.NaN);
            Assert.That(v.F64Inf, Is.EqualTo(double.PositiveInfinity));
            Assert.That(v.F64MinusInf, Is.EqualTo(double.NegativeInfinity));
            
            Assert.That(v.ListValue, Is.EquivalentTo((VList<int>)[ 1, 2, 3 ]));
            
            Assert.That(v.OptionalFieldSome.Field.TryGetValue(out var value) && value == 4, Is.True);
            Assert.That(v.OptionalFieldNone.Field.IsSome, Is.False);
            
            Assert.That(v.DictValue, Is.EquivalentTo(new Dictionary<string, int> { ["a"] = 1, ["b"] = 2 }));
            Assert.That(v.DictFieldValue.Field, Is.EquivalentTo(new Dictionary<string, int> { ["a"] = 1, ["b"] = 2 }));
        });
    }
    
}

File: langs/java/test/src/test/java/dev/argon/nobleidl/test/tests/DefaultValueTests.java
package dev.argon.nobleidl.test.tests;


import dev.argon.esexpr.ESExpr;
import dev.argon.nobleidl.example.DefaultValues;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.math.BigInteger;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class DefaultValueTests {

	@Test
	public void checkDefaultValues() throws Exception {
		var v = DefaultValues.codec().decode(new ESExpr.Constructor("default-values", List.of(), Map.of()));

		assertTrue(v.boolTrue());
		assertFalse(v.boolFalse());
		assertEquals("abc", v.strValue());
		assertArrayEquals(new byte[] { (byte)0xAB, (byte)0xCD, (byte)0xEF }, v.binaryValue());

		assertEquals(BigInteger.valueOf(55), v.intValue55());
		assertEquals(new BigInteger("18446744073709551616"), v.intBeyondInt64());
		assertEquals(BigInteger.valueOf(-55), v.intMinus55());
		assertEquals(new BigInteger("-18446744073709551617"), v.intBeyondNegInt64());
		assertEquals(BigInteger.valueOf(55), v.natValue55());
		assertEquals(new BigInteger("18446744073709551616"), v.natBeyondInt64());

		assertEquals((byte)55, v.i8Value55());
		assertEquals((byte)-55, v.i8Minus55());
		assertEquals((byte)55, v.u8Value55());
		assertEquals((byte)255, v.u8Value255());
		assertEquals((short)55, v.i16Value55());
		assertEquals((short)-55, v.i16Minus55());
		assertEquals((short)55, v.u16Value55());
		assertEquals((short)65535, v.u16ValueMax());
		assertEquals(55, v.i32Value55());
		assertEquals(-55, v.i32Minus55());
		assertEquals(55, v.u32Value55());
		assertEquals((int)4294967295L, v.u32ValueMax());
		assertEquals(55L, v.i64Value55());
		assertEquals(-55L, v.i64Minus55());
		assertEquals(55L, v.u64Value55());
		assertEquals(new BigInteger("18446744073709551615").longValue(), v.u32ValueMax());

		assertEquals(55.0f, v.f32Value55());
		assertEquals(-55.0f, v.f32Minus55());
		assertTrue(Float.isNaN(v.f32Nan()));
		assertEquals(Float.POSITIVE_INFINITY, v.f32Inf());
		assertEquals(Float.NEGATIVE_INFINITY, v.f32MinusInf());

		assertEquals(55.0, v.f64Value55());
		assertEquals(-55.0, v.f64Minus55());
		assertTrue(Double.isNaN(v.f32Nan()));
		assertEquals(Double.POSITIVE_INFINITY, v.f64Inf());
		assertEquals(Double.NEGATIVE_INFINITY, v.f64MinusInf());

		assertEquals(List.of(1, 2, 3), v.listValue());

		assertEquals(Optional.of(4), v.optionSome());
		assertEquals(Optional.empty(), v.optionNone());
		assertEquals(Optional.of(Optional.of(4)), v.option2SomeSome());
		assertEquals(Optional.of(Optional.empty()), v.option2SomeNone());
		assertEquals(Optional.empty(), v.option2None());

		assertEquals(Optional.of(4), v.optionalFieldSome().field());
		assertEquals(Optional.empty(), v.optionalFieldNone().field());

		assertEquals(Map.of("a", 1, "b", 2), v.dictValue().map());
		assertEquals(Map.of("a", 1, "b", 2), v.dictFieldValue().field().map());
	}

}
